File: Level_3/광고삽입/ad.js
function solution(play_time, adv_time, logs) {
	// 시간을 보다 쉽게 다루기 위해서 정수형으로 변환한다.
	const intPlayTime = time2int(play_time);
	const intAdvTime = time2int(adv_time);
	// advTime 이 playTime 보다 긴 경우, 처음부터 트는 것이 가장 합당하다.
	if (intPlayTime <= intAdvTime) return "00:00:00";

	// 각 시작, 끝 시간에 대해서 조사를 하는 것이 바람직하다.
	const timeSet = new Set();

	// log에 포함된 시간값 또한, 정수로 변환하여 다룬다.
	const l = [];
	const REG = RegExp("(?<s_time>[0-9]{2}:[0-9]{2}:[0-9]{2})-(?<e_time>[0-9]{2}:[0-9]{2}:[0-9]{2})");
	logs.forEach((elem) => {
		const t = elem.match(REG).groups;
		const [sTime, eTime] = [time2int(t.s_time), time2int(t.e_time)];
		l.push([sTime, eTime]);
		timeSet.add(sTime);
	});
	const intLogs = l.sort((a, b) => {
		if (a[0] < b[0]) {
			return -1;
		} else if (a[0] === b[0]) {
			return 0;
		} else {
			return 1;
		}
	});
	// time  ~ time + advTime 내에 해당 재생 시간이 포함되면 누적 시간에 포함함.
	const checkTimes = Array.from(timeSet);
	let max = -1;
	let intResult = 0;
	checkTimes.forEach((checkTime) => {
		const cStartTime = +checkTime;
		const cEndTime = +checkTime + intAdvTime;
		let accTime = 0;
		for (const [sTime, eTime] of intLogs) {
			if (cEndTime < sTime) break;
			else if (eTime < cStartTime) continue;
			if (cStartTime <= sTime && eTime <= cEndTime) {
				accTime += eTime - sTime;
			} else if (cStartTime <= sTime && cEndTime <= eTime && sTime <= cEndTime) {
				accTime += cEndTime - sTime;
			} else if (sTime <= cStartTime && eTime <= cEndTime && cStartTime <= eTime) {
				accTime += eTime - cStartTime;
			}
		}
		if (max < accTime) {
			max = accTime;
			intResult = checkTime;
		} else if (max === accTime) {
			intResult = intResult > checkTime ? checkTime : intResult;
		}
	});
	// console.log(int2time(intResult),int2time(max));
	return int2time(intResult);
}

function time2int(time) {
	const REG = RegExp("(?<h>[0-9]{2}):(?<m>[0-9]{2}):(?<s>[0-9]{2})");
	const times = time.match(REG).groups;
	const result = times.h * 3600 + times.m * 60 + +times.s;
	return +result;
}
function int2time(n) {
	const h = Math.floor(n / 3600);
	const m = Math.floor((n - h * 3600) / 60);
	const s = n - h * 3600 - m * 60;
	return `${addZero(h)}:${addZero(m)}:${addZero(s)}`;
}
function addZero(n) {
	const str = n.toString();
	return str.length <= 1 ? "0" + str : str;
}

// play time 내에서 검색
// play time -> int로 변환하자.
// h*3600 + m *60 + s
// log 에서 s_time, e_time으로 나눠서 각각 저장함.
// 추석 트래픽에서 적용한 알고리즘과 유사하게 적용하거나, sliding window를 적용하는 것이 바람직하다.
// window의 크기는 adv_time

File: Level_3/광고삽입/ad.cpp
#include <string>
#include <vector>
#include <regex>
#include <iostream>
#include <set>
#include <map>

using namespace std;


long long time2int(string time){
    regex REG("([0-9]{2}):([0-9]{2}):([0-9]{2})");
    smatch matches;
    int h,m,s;
    if(regex_search(time,matches,REG)){
        h = stoi(matches[1].str());
        m = stoi(matches[2].str());
        s = stoi(matches[3].str());
    } 
    return h*3600 + m*60 + s;
}

string addZero(long long n){
    if(n<10){
        return "0"+to_string(n);
    } else {
        return to_string(n);
    }
}

string int2time(long long n){
    const int h = n/3600;
    const int m = (n - 3600*h)/60;
    const int s = (n - 3600*h - 60*m);
    return addZero(h)+":"+addZero(m)+":"+addZero(s);
}

string solution(string play_time, string adv_time, vector<string> logs) {
    string answer = "";
    long long lPlayTime = time2int(play_time);
    long long lAdvTime = time2int(adv_time);
    if(lPlayTime <= lAdvTime) return "00:00:00";

    set<long long> timeSet;
    map<long long, int> timeMap;
    vector<pair<long long, long long>> intLogs;

    regex rex("([0-9]{2}:[0-9]{2}:[0-9]{2})-([0-9]{2}:[0-9]{2}:[0-9]{2})");
    for(const string log : logs){
        smatch matches;
        if(regex_search(log, matches, rex)){
            const long long sTime = time2int(matches[1].str());
            const long long eTime = time2int(matches[2].str());
            timeSet.insert(sTime);
            timeSet.insert(eTime);
            intLogs.push_back({sTime,eTime});
        }
    }
    int i = 0;
    vector<long long> temp(timeSet.begin(), timeSet.end());
    vector<vector<long long>> bar;
    for(int i=0; i<temp.size(); i++){
        // time, count, duration
        const int duration = i<(temp.size()-1)? temp[i+1] - temp[i] : 0;
        bar.push_back({temp[i],0,duration});
        timeMap.insert({temp[i],i});
    }
    for(auto p : intLogs){
        auto sTime = p.first;
        auto eTime = p.second;
        for(int i = timeMap[sTime]; bar[i][0] < eTime; i++){
            if (sTime <= bar[i][0] && bar[i][0] < eTime) {
				bar[i][1]++;
			}
        }
    }
    
    long long max = 0;
    long long res = 0;
    for(int i = 0 ; i<bar.size() -1; i++){
        const long long sTime = bar[i][0];
        const long long eTime = sTime + lAdvTime;
        long long acc = 0;
        for(int j=i; j<bar.size(); j++){
            const long long time = bar[j][0];
            const long long count = bar[j][1];
            const long long duration = bar[j][2];
            if(time > eTime) break;
            if(time + duration <= eTime){
                acc += count * duration;
            } else {
                acc += (eTime - time) * count;
            }
        }
        if( max < acc ){
            max = acc;
            res = sTime;
        }
    }
    answer = int2time(res);
    return answer;
}

